apps/ex_wire/lib/ex_wire/dev_p2p.ex

```diff
@@ -0,0 +1,156 @@
+    def disconnect(session = %Session{}) do
+      %{session | handshake_sent: false, handshake_received: false}
+    end
+
+    @spec compatible_capabilities?(t) :: boolean()
+    def compatible_capabilities?(session = %Session{}) do
+      %Session{handshake_received: handshake_received, handshake_sent: handshake_sent} = session
+
+      intersection =
+        MapSet.intersection(
+          to_mapset(handshake_received.caps),
+          to_mapset(handshake_sent.caps)
+        )
+
+      !Enum.empty?(intersection)
+    end
+
+    defp to_mapset(list) do
+      Enum.into(list, MapSet.new())
+    end
+  end
+
+  @doc """
+  Convenience function to create an `ExWire.DEVp2p.Session` struct
+  """
+  @spec init_session :: Session.t()
+  def init_session do
+    %Session{}
+  end
+
+  @doc """
+  Function to create a DEVp2p struct needed for a protocol handshake. This
+  should be an `ExWire.Packet.Hello` struct with the appropriate values filled in.
+  """
+  @spec generate_handshake :: Packet.Hello.t()
+  def generate_handshake do
+    %Packet.Hello{
+      p2p_version: Config.p2p_version(),
+      client_id: Config.client_id(),
+      caps: Config.caps(),
+      listen_port: Config.listen_port(),
+      node_id: Config.node_id()
+    }
+  end
+
+  @doc """
+  Function to update `ExWire.DEVp2p.Session` when a handshake is sent. The
+  handshake should be an `ExWire.Packet.Hello` that we have sent to a peer.
+  """
+  @spec handshake_sent(Session.t(), Packet.Hello.t()) :: Session.t()
+  def handshake_sent(session, hello = %Packet.Hello{}) do
+    %{session | handshake_sent: hello}
+  end
+
+  @doc """
+  Function to update `ExWire.DEVp2p.Session` when a handshake is received. The
+  handshake should be an `ExWire.Packet.Hello` that we have received from a peer.
+  """
+  @spec handshake_received(Session.t(), Packet.Hello.t()) :: Session.t()
+  def handshake_received(session, hello = %Packet.Hello{}) do
+    %{session | handshake_received: hello}
+  end
+
+  @doc """
+  Function to check whether or not a `ExWire.DEVp2p.Session` is active. See
+  `ExWire.DEVp2p.Session.active?/1` for more information.
+  """
+  @spec session_active?(Session.t()) :: boolean()
+  def session_active?(session), do: Session.active?(session)
+
+  @spec session_compatible?(Session.t()) :: boolean()
+  def session_compatible?(session), do: Session.compatible_capabilities?(session)
+
+  @doc """
+  Function to handles other messages related to the DEVp2p protocol that a peer
+  sends. The messages could be `ExWire.Packet.Disconnect`, `ExWire.Packet.Ping`,
+  or `ExWire.Packet.Pong`.
+
+  An `ExWire.DEVp2p.Session` is required as the first argument in order to
+  properly update the session based on the message received.
+  """
+  @spec handle_message(Session.t(), struct()) ::
+          {:error, :handshake_incomplete, Session.t()}
+          | {:ok, ExWire.Pakcet.handle_response(), Session.t()}
+  def handle_message(session, message) do
+    case session_active?(session) do
```

Comment:
why not `if session_active?(session) do`?

Reply:
No particular reason. It could totally work as an `if, else`. I think I'm just used to thinking in `case` statements and pattern matching in elixir. 

I'll update this to use `if else`. I think it makes more sense.